resources/testharness.js

```diff
@@ -2979,11 +2985,30 @@ policies and contribution forms [3].
     /*
      * Utility functions
      */
+
+    let continue_execution_after_failing_assertion = false;
+    function execute_single_page_test_after_failing_assertion()
+    {
+        continue_execution_after_failing_assertion = true;
+    }
+    expose(execute_single_page_test_after_failing_assertion, 'execute_single_page_test_after_failing_assertion');
+
     function assert(expected_true, function_name, description, error, substitutions)
     {
         if (tests.tests.length === 0) {
             tests.set_file_is_test();
         }
+        if (tests.file_is_test) {
+            let test = tests.tests[0];
+            let message = expected_true ? description : make_message(function_name, description, error, substitutions);
+            test.add_assertion({passed: !!expected_true, message: message});
+            if (continue_execution_after_failing_assertion) {
+                if (!expected_true && test.status !== test.FAIL) {
+                    test.set_status(test.FAIL, "At least one assertion failed: " + message);
```

Comment:
Looks like this will list the first failure. Is the idea that the treatment of `assertions` is entirely vendor-specific, or how will one make use of the list of assertions?

For wpt.fyi, how should this feature change what is shown? Since this is limited to single-page tests we *could* try to treat the assertions as tests in themselves, but then we'll also have to require that each assertion's message is unique, or possibly make it unique automatically if it isn't. The only other option I see is to require the assertion order to be stable across runs or between browsers, as otherwise there's no way to compare results.

@lukebjerring for more ideas.

Reply:
My 5c: subtests are different that assertions, which are more like a precondition (such as asserting that something is non-null, before asserting the correct value of a property on it).
If you want to have separate failures for multiple things, that don't cascade, then we should be creating subtests. 
Something like https://github.com/lukebjerring/wpt/pull/14 is what I mean in this specific case.

Reply:
I agree with @lukebjerring here; this model looks like it should work by wrapping each `assert()` in an implicit `test()`;

Reply:
> Looks like this will list the first failure. Is the idea that the treatment of `assertions` is entirely vendor-specific, or how will one make use of the list of assertions?

Yes, this would be vendor specific.
WebKittens want more reporting in what is produced by WebKitTestRunner.
Ideally, for every failed test (single file test or not), it would be good to have the list of assertions that passed for the given test.
testharnessreport.js could be used to display that information in whatever form.
For instance, it could be dumped in stderr for debugging purposes by WebKitTestRunner and on the inspector console on a browser.

The current PR provides the first error message in the test failing status.
To decrease flakiness, it might just be better to not report it and let testharnessreport.js actually decide how it wants to output the information.

Implicit test() is another valid approach.
My main concern is how much flakiness this actually ends up creating in the output.
Test writers will need to ensure that assertions are called the exact same amount of time in the exact same order. I am not sure though how practical this is to require in WPT.
WebKit js-test actually requires that so I am fine trying it here as well.

Reply:
The order shouldn't be important; it's not part of the API. You may need to find a way to create a unique identity for the alert to ensure the order doesn't leak.